### DISCUSSION TITLE
Validate always and pre

### DISCUSSION THREAD

**Discussion by frankie567** [pydantic/pydantic#2640] [Question] [answered]
Hi there!

I have a question regarding the combination of the `always` and `pre` arguments of the `validator` decorator. It seems more or less related to #2376 and #1801.

In [the documentation](https://pydantic-docs.helpmanual.io/usage/validators/#validate-always), this example is shown:

```py
from datetime import datetime

from pydantic import BaseModel, validator


class DemoModel(BaseModel):
    ts: datetime = None

    @validator('ts', pre=True, always=True)
    def set_ts_now(cls, v):
        return v or datetime.now()


print(DemoModel())
#> ts=datetime.datetime(2021, 3, 3, 17, 30, 4, 413655)
print(DemoModel(ts='2017-11-08T14:00'))
#> ts=datetime.datetime(2017, 11, 8, 14, 0)
```

`mypy` complains here, but that is expected, and the example works perfectly well. Below, the documentation explains this:

> You'll often want to use this together with pre, since otherwise with always=True pydantic would try to validate the default None which would cause an error.

However, I found that this doesn't seem to be true, at least for this example. If we are removing `pre=True`, it works also flawlessly:

```py
from datetime import datetime

from pydantic import BaseModel, validator


class DemoModel(BaseModel):
    ts: datetime = None

    @validator('ts', always=True)
    def set_ts_now(cls, v):
        return v or datetime.now()


print(DemoModel())
#> ts=datetime.datetime(2021, 3, 3, 17, 30, 4, 413655)
print(DemoModel(ts='2017-11-08T14:00'))
#> ts=datetime.datetime(2017, 11, 8, 14, 0)
```

My understanding is that since we provide a default value, Pydantic is perfectly fine with it and doesn't trigger a validation error. It's even more true if we write `ts: Optional[datetime] = None` to be strictly compliant with `mypy`.

So my question is this: **does `pre=True` is actually needed when we want to set a dynamic default value?**

## Version info

```
pydantic version: 1.8.1
pydantic compiled: False
python version: 3.7.10 (default, Mar 29 2021, 14:42:56)  [Clang 12.0.0 (clang-1200.0.32.29)]
platform: Darwin-20.3.0-x86_64-i386-64bit
optional deps. installed: ['dotenv', 'email-validator', 'typing-extensions']
```

**Comment by PrettyWood** [comment, marked as answer]
Hi @frankie567 
Setting a default value makes the field optional (vs required).
`None` is a special default value that changes the type `datetime` into `Optional[datetime]` (it will probably [be changed in v2](https://github.com/samuelcolvin/pydantic/discussions/2353)) so in this example it's true that `pre` does not really makes sense since `None` will still be parsed as `None`
But if you change the default value
```py
class DemoModel(BaseModel):
    ts: datetime = 0
```
you'll have `v=0` if you use `pre=True` and `v=datetime.datetime(1970, 1, 1, 0, 0, tzinfo=datetime.timezone.utc)` without (parsing has been done!)

----------------------------------------------
You have many ways to create a dynamic value

#### `default_factory`
```py
from datetime import datetime

from pydantic import BaseModel, Field


class DemoModel(BaseModel):
    ts: datetime = Field(default_factory=datetime.now)
```

#### `root_validator`
```py
from datetime import datetime

from pydantic import BaseModel, root_validator


class DemoModel(BaseModel):
    ts: datetime

    @root_validator(pre=True)
    def set_ts_now(cls, values: dict[str, object]) -> dict[str, object]:
        values.setdefault('ts', datetime.now())
        return values
```
(the `root_validator(pre=True)` is run before everything)

#### `validator`
```py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field, validator


# https://www.python.org/dev/peps/pep-0484/#id30
class Unset(Enum):
    token = 0

_unset = Unset.token


class DemoModel(BaseModel):
    ts: datetime = Field(_unset)

    @validator('ts', pre=True, always=True)
    def set_ts_now(cls, v: datetime | Unset) -> datetime:
        if v is _unset:
            return datetime.now()
        return v
```

Hope that clarifies things

**Reply by frankie567** [reply]
It does, thanks! The documentation sounds a bit misleading here: `pre` will determine if we get a parsed value or not, but clearly won't cause an error if we omit it, right?

Obviously, the `default_factory` argument looks like the best way to go; but I was a little worried about its *beta* status.

The `Unset` token trick is interesting, I also came with this idea at some point 😄

**Reply by PrettyWood** [reply]
`pre` just means "validate the raw value before it has been parsed/coerced", `always` means "do this also for set default values".
If a field is required and no value (or default value) has been set it will crash.
`default_factory` works well and has been in beta since 1.5! It's reliable and we shall remove the 'beta' mention for v2 ;)

**Reply by frankie567** [reply]
Yes, that's crystal clear 👍

But at the light of this, I tend to think that this [part of the documentation after the example](https://pydantic-docs.helpmanual.io/usage/validators/#validate-always) should be rephrased a bit 🙃

**Reply by PrettyWood** [reply]
Totally agree. PR welcome ;D
